File: backend/src/main/java/com/restaurant/service/reservation/ReservationService.java
package com.restaurant.service.reservation;

import com.restaurant.common.enums.ReservationStatus;
import com.restaurant.dto.request.reservation.CreateReservationRequest;
import com.restaurant.dto.request.reservation.UpdateReservationRequest;
import com.restaurant.dto.response.reservation.BookingSuggestionResponse;
import com.restaurant.dto.response.reservation.ReservationCalendarResponse;
import com.restaurant.dto.response.reservation.ReservationResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

public interface ReservationService {

    // Lấy danh sách lịch đặt bàn phân trang (hỗ trợ lọc linh hoạt theo ngày và/hoặc trạng thái)
    Page<ReservationResponse> getReservations(LocalDate date, ReservationStatus status, Pageable pageable);

    // Lấy thông tin chi tiết một lượt đặt bàn theo ID
    ReservationResponse getById(UUID id);

    // Tiếp nhận thông tin đặt bàn mới, lưu ở trạng thái PENDING; capacity check được thực hiện ở bước confirmReservation
    ReservationResponse createReservation(CreateReservationRequest request, UUID staffId);

    // Xác nhận đơn PENDING: kiểm tra capacity → CONFIRMED nếu còn chỗ, tự động CANCELLED nếu hết chỗ
    ReservationResponse confirmReservation(UUID id, UUID staffId);

    // Cập nhật thông tin khách hàng hoặc thời gian của một đơn đặt bàn (cho phép sửa ở trạng thái PENDING/CONFIRMED)
    ReservationResponse update(UUID id, UpdateReservationRequest request);

    // Đánh dấu khách đã đến nhận bàn; đồng thời chuyển trạng thái bàn được gán sang SERVING
    ReservationResponse arrived(UUID id);

    // Huỷ đơn đặt bàn; giải phóng bàn (nếu có) và lưu vết người huỷ cùng lý do
    ReservationResponse cancel(UUID id, UUID cancelledBy, String reason);

    // Tác vụ tự động: quét và gán bàn trống tối ưu cho các đơn CONFIRMED sắp đến giờ
    void autoAssignTables();

    // Tác vụ tự động: huỷ các đơn CONFIRMED quá giờ hẹn mà khách không check-in
    void autoCancelExpired();

    // Lấy thông tin tổng hợp đặt bàn trong ngày, bóc tách theo từng trạng thái (dùng cho Dashboard/Calendar)
    ReservationCalendarResponse getCalendar(LocalDate date);

    // Gợi ý các khung giờ còn trống theo danh sách ngày và số người khách cung cấp
    List<BookingSuggestionResponse> suggestBookingSlots(List<LocalDate> preferredDates, Integer partySize);

    void delete(UUID id);
}

File: frontend-admin/src/services/reservation.service.ts
// frontend/src/services/reservation.service.ts
import { apiClient } from "@/lib/api-client";
import { ApiResponse, PaginatedData } from "@/types/common";
import {
  Reservation,
  ReservationStatus,
  ReservationCalendar,
  CreateReservationRequest,
  UpdateReservationRequest,
  CancelReservationRequest,
} from "@/types/reservation";

export const reservationService = {
  /**
   * GET /api/reservations
   * Lấy danh sách đặt bàn, hỗ trợ filter theo ngày và trạng thái
   */
  async getReservations(params?: {
    date?: string;
    status?: ReservationStatus;
    page?: number;
    size?: number;
  }): Promise<PaginatedData<Reservation>> {
    const query = new URLSearchParams();
    if (params?.date) query.append("date", params.date);
    if (params?.status) query.append("status", params.status);
    if (params?.page !== undefined) query.append("page", String(params.page));
    if (params?.size !== undefined) query.append("size", String(params.size));
    const queryStr = query.toString();
    const endpoint = queryStr ? `/reservations?${queryStr}` : `/reservations`;

    // Controller trả về Page<ReservationResponse> trực tiếp (không bọc ApiResponse)
    return apiClient<PaginatedData<Reservation>>(endpoint, {
      method: "GET",
      requireAuth: true,
    });
  },

  /**
   * GET /api/reservations/calendar?date={date}
   * Lấy lịch đặt bàn theo ngày (tổng hợp số lượng theo trạng thái)
   */
  async getCalendar(date: string): Promise<ReservationCalendar> {
    // Controller trả về ReservationCalendarResponse trực tiếp (không bọc ApiResponse)
    return apiClient<ReservationCalendar>(`/reservations/calendar?date=${date}`, {
      method: "GET",
      requireAuth: true,
    });
  },

  /**
   * GET /api/reservations/{id}
   * Lấy chi tiết một đặt bàn theo id
   */
  async getReservationById(id: string): Promise<Reservation> {
    // Controller trả về ReservationResponse trực tiếp
    return apiClient<Reservation>(`/reservations/${id}`, {
      method: "GET",
      requireAuth: true,
    });
  },

  /**
   * POST /api/reservations
   * Tạo đặt bàn mới (có thể kèm X-Staff-ID header nếu nhân viên tạo)
   */
  async createReservation(data: CreateReservationRequest, staffId?: string): Promise<Reservation> {
    const headers: Record<string, string> = {};
    if (staffId) headers["X-Staff-ID"] = staffId;

    return apiClient<Reservation>("/reservations", {
      method: "POST",
      body: JSON.stringify(data),
      headers,
      requireAuth: true,
    });
  },

  /**
   * PUT /api/reservations/{id}
   * Cập nhật thông tin đặt bàn
   */
  async updateReservation(id: string, data: UpdateReservationRequest): Promise<Reservation> {
    return apiClient<Reservation>(`/reservations/${id}`, {
      method: "PUT",
      body: JSON.stringify(data),
      requireAuth: true,
    });
  },

  /**
   * PUT /api/reservations/{id}/confirm
   * Xác nhận đặt bàn (kèm X-Staff-ID header)
   */
  async confirmReservation(id: string, staffId: string): Promise<Reservation> {
    return apiClient<Reservation>(`/reservations/${id}/confirm`, {
      method: "PUT",
      headers: { "X-Staff-ID": staffId },
      requireAuth: true,
    });
  },

  /**
   * PUT /api/reservations/{id}/arrived
   * Đánh dấu khách đã đến
   */
  async markAsArrived(id: string): Promise<Reservation> {
    return apiClient<Reservation>(`/reservations/${id}/arrived`, {
      method: "PUT",
      requireAuth: true,
    });
  },

  /**
   * PUT /api/reservations/{id}/no-show
   * Đánh dấu khách không đến
   */
  async markAsNoShow(id: string): Promise<Reservation> {
    return apiClient<Reservation>(`/reservations/${id}/no-show`, {
      method: "PUT",
      requireAuth: true,
    });
  },

  /**
   * PUT /api/reservations/{id}/cancel
   * Hủy đặt bàn kèm lý do (có thể kèm X-Staff-ID header)
   */
  async cancelReservation(
    id: string,
    data: CancelReservationRequest,
    staffId?: string
  ): Promise<Reservation> {
    const headers: Record<string, string> = {};
    if (staffId) headers["X-Staff-ID"] = staffId;

    return apiClient<Reservation>(`/reservations/${id}/cancel`, {
      method: "PUT",
      body: JSON.stringify(data),
      headers,
      requireAuth: true,
    });
  },

  /**
   * GET /api/reservations/available-slots?date={date}&partySize={partySize}
   * Lấy danh sách giờ còn trống trong ngày cho số lượng khách
   */
  async getAvailableSlots(date: string, partySize: number): Promise<string[]> {
    const query = new URLSearchParams({ date, partySize: String(partySize) }).toString();
    // Controller trả về List<LocalTime> trực tiếp
    return apiClient<string[]>(`/reservations/available-slots?${query}`, {
      method: "GET",
      requireAuth: true,
    });
  },

  /**
   * DELETE /api/reservations/{id}
   * Xóa đặt bàn
   */
  async deleteReservation(id: string): Promise<void> {
    return apiClient<void>(`/reservations/${id}`, {
      method: "DELETE",
      requireAuth: true,
    });
  },
};
